### PR TITLE
Test: add testing for adb commands

### PR DESCRIPTION
**Pull request by andrewqiamzn**
### Why?
The DAB-ADB bridge currently has no tests. We want to add unit tests with Jest and achieve minimum 80% coverage.
### DOD
- [ ] Achieve 80% test coverage `adb_commands.ts`
### Test Plan
image

### CODE REVIEW

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 1]
```diff
@@ -0,0 +1,969 @@
...
+    beforeEach(() => {
+        mockSpawn = jest.fn(() => Promise.resolve());
+
+        asMock(spawn).mockImplementation(mockSpawn);
+        asMock(config.get).mockImplementation(mockConfigGetAdbBinary);
+
+        testAdbCommands = new AdbCommands(testDeviceId);
+    });
+
+    describe("getDeviceDetails", () => {
```
Comment (andrewqiamzn) [inline adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:61]: This function was by far the hardest to test. Calling `devices()` would lead to a huge number of different code flows due to its nested loops and if statements.  Subsequent private function calls such as `getDeviceProps()`, `getNetworkType()`, `getDeviceIpsFromSerial()`, `getDeviceMacFromIp()` would involve regex parsing spawn outputs, which was difficult to test due to the lack of documentation on responses to adb commands. 

Suggested future work ->
- [ ] Modularise out regex sections into its own class, so these can be tested separately. Adjust testing accordingly
- [ ] Research and note down documentation on expected adb command responses
- [ ] Increase branch coverage for this file to 80%

**Review by andrewqiamzn — COMMENTED** [review]
It was becoming increasingly difficult and in-scalable to test [getDevicesDetails()](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L93). 

Calling the private function [devices()](https://github.com/device-automation-bus/dab-tools/blob/main/adapters/dab-adb-bridge/src/lib/adb/adb_commands.ts#L157) would lead to a huge number of different code flows due to its nested loops and if statements.  

Subsequent private function calls such as `getDeviceProps()`, `getNetworkType()`, `getDeviceIpsFromSerial()`, `getDeviceMacFromIp()` would involve regex parsing for `spawn` outputs, which was difficult to test due to the lack of documentation on these outputs. Depending on the spawn output and how it matches with regex, additional code flows would need to be tested. Before spending more time on testing these additional code flows, I recommend it would be better modularise/refactor certain parts of the code to increase readability and ease of testing.

Suggested future work ->
- [ ] Modularise out regex sections into its own class, so these can be tested separately. Adjust testing in this file accordingly
- [ ] Research and document expected adb command responses
- [ ] Increase branch coverage for this file to 80%
- [ ] Refactor and reduce complexity of `devices()` function. Extract out to util functions when possible.

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 2]
```diff
@@ -0,0 +1,969 @@
...
+            await expect(testAdbCommands.reboot()).resolves.not.toThrow();
+        });
+        it("should reject and throw an error if timeout is exceeded", async () => {
+            jest.useFakeTimers();
+            const expectedRebootError = "Reconnection timeout following reboot";
+
+            const mockRebootSpawn = jest.fn(() => Promise.resolve());
+            const mockConfigGetAdbRebootTimeoutMs = jest.fn(() => 100);
+            asMock(spawn).mockImplementationOnce(mockRebootSpawn);
+            asMock(config.get).mockImplementation(mockConfigGetAdbRebootTimeoutMs);
```
Comment (giotar) [inline adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:915]: As before, you can write `asMock(config.get).mockResolvedValue(100)`

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 3]
```diff
@@ -0,0 +1,969 @@
...
+        testAdbCommands = new AdbCommands(testDeviceId);
+    });
+
+    describe("getDeviceDetails", () => {
+        describe("when device is using a usb connection", () => {
+            it("should throw an error if spawn fails", async () => {
+                const expectedGetDeviceDetailsError = `Spawn has failed`;
+
+                mockSpawn = jest.fn(() => Promise.reject(new Error(expectedGetDeviceDetailsError)));
+                asMock(spawn).mockImplementation(mockSpawn);
```
[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:67]: Thanks! Really good point, will improve that in my next commit!

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 4]
```diff
@@ -0,0 +1,969 @@
...
+                    Promise.resolve({
+                        stdout: Buffer.from(`connected to ${testDeviceIpAddress}`, "utf8"),
+                        stderr: stderrBufferWithoutError
+                    })
+                );
+
+                asMock(spawn).mockImplementationOnce(mockSpawnFromDevice);
+                asMock(spawn).mockImplementationOnce(mockSpawnFromConnect);
+                asMock(spawn).mockImplementationOnce(mockSpawnFromDevice);
+                asMock(spawn).mockImplementationOnce(mockSpawnFromDevice);
```
[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:246]: Very true, implementing a dynamic mock would be much cleaner and then i can also check how many times it has been called too!

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 5]
```diff
@@ -0,0 +1,969 @@
...
+                    Promise.resolve({
+                        stdout: Buffer.from(`${testDeviceIpAddressWithColon} device`, "utf8"),
+                        stderr: stderrBufferWithoutError
+                    })
+                );
+
+                asMock(spawn).mockImplementationOnce(mockSpawn);
+                asMock(spawn).mockImplementationOnce(mockSpawn);
+                asMock(spawn).mockImplementationOnce(mockSpawn);
+                asMock(spawn).mockImplementationOnce(mockSpawn);
```
[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:378]: In this code flow, spawn would be called four times, so any less than 4 then it would trigger this error
Screenshot 2022-12-15 at 13 44 00

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 6]
```diff
@@ -0,0 +1,969 @@
...
+        `("should start the intent and return if $message", async ({ message }) => {
+            const mockSpawn = jest.fn(() =>
+                Promise.resolve({
+                    stdout: stdoutBufferWithoutError,
+                    stderr: Buffer.from(message, "utf8")
+                })
+            );
+            asMock(spawn).mockImplementationOnce(mockSpawn);
+
+            await expect(testAdbCommands.start(intentArr)).resolves.not.toThrow();
```
[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:525]: Oh I see! Will update in my next commit thanks for the tip

File: adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts [hunk 7]
```diff
@@ -0,0 +1,969 @@
...
+
+        it("should switch to call status2 if calling status1 fails (could not read stdout/stderr)", async () => {
+            const expectedStatusResponse = { package: testAppPackage, state: AndroidApplicationStatus.Stopped };
+
+            const status1MockSpawn = jest.fn(() => Promise.resolve(nullProcessOutputObject));
+            const status2MockSpawn = jest.fn(() =>
+                Promise.resolve({ stdout: stdoutBufferWithoutError, stderr: stderrBufferWithoutError })
+            );
+            asMock(spawn).mockImplementationOnce(status1MockSpawn);
+            asMock(spawn).mockImplementationOnce(status2MockSpawn);
```
[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:755]: Got it, will just use a single mock. Should I create a single mock for ALL the different mock calls or separate it out by each `describe` section?

[the remark this answers is not in the document]
Comment (andrewqiamzn) [reply adapters/dab-adb-bridge/src/lib/adb/__tests__/adb_commands.test.ts:755]: I think seperating it out would be better as otherwise it would be a huge list of if statements! 😆

**Review by giotar — APPROVED** [review]
